simulation/amaru-sim/src/simulator/run.rs

```diff
@@ -240,39 +237,112 @@ fn make_chain_selector(
     SelectChain::new(tree_state)
 }
 
-/// Property: at the end of the simulation, the tip of the chain from the last block must
-/// match the tip of the chain returned by the history of messages sent by the nodes under test.
-fn chain_property(
-    chain_data_path: &PathBuf,
-) -> impl Fn(&History<ChainSyncMessage>) -> Result<(), String> + use<'_> {
-    move |history| {
-        // Get the tip of the chain from the chain data file
-        // and compare it to the last entry in the history
-        let blocks =
-            Chain::read_blocks_from_file(chain_data_path).map_err(|err| err.to_string())?;
-        let expected = blocks
-            .last()
-            .map(|block| (block.hash.clone(), Slot::from(block.slot)))
-            .expect("empty chain data");
+/// Property: at the end of the simulation, the chain built from the history of messages received
+/// downstream must match the best chain built directly from messages coming from upstream peers.
+fn chain_property()
+-> impl Fn(&[Entry<ChainSyncMessage>], &History<ChainSyncMessage>) -> Result<(), String> {
+    move |entries, history| {
+        let expected = make_best_chain_from_upstream_messages(
+            &entries
+                .iter()
+                .map(|e| e.envelope.clone())
+                .collect::<Vec<_>>(),
+        );
+        let actual = make_best_chain_from_downstream_messages(history);
+        let actual_chain = actual.list_to_string(",\n ");
+        let expected_chain = expected.list_to_string(",\n ");
+        assert_eq!(
+            actual, expected,
+            "\nThe actual chain\n{}\n\nis not the best chain\n\n{}\n\nThe history is:\n{:?}",
+            actual_chain, expected_chain, history,
+        );
+        Ok(())
+    }
+}
+
+/// Build the best chain from messages incoming from upstream peers.
+pub fn make_best_chain_from_upstream_messages(messages: &[Envelope<ChainSyncMessage>]) -> Chain {
```

Comment:
It seems odd we have to write this complicated logic to compute the best chain from upstream messages, are we not duplicating some of the logic of the chain selection? Shouldn't the chain generator already know what the best chain is? I think we already had that discussion 😓

Reply:
This is simpler logic than chain selection but you're right we can keep the original spine for the generated tree of headers and pass it around. I pushed a commit for this.